src/pstl_offload.cpp

```diff
@@ -130,32 +147,254 @@ __internal_free(void* __user_ptr)
+        }
+    }
+}
+
+static void
+__internal_free(void* __user_ptr)
+{
+    __internal_free_param(__user_ptr, __original_free_ptr);
+}
+
+#endif // _WIN64
+
 static std::size_t
 __internal_msize(void* __user_ptr)
 {
+    if (__user_ptr == nullptr)
+    {
+#if _WIN64
+        errno = EINVAL;
+        return -1;
+#else
+        return 0;
+#endif
+    }
+
+    std::size_t __res = 0;
+    __block_header* __header = static_cast<__block_header*>(__user_ptr) - 1;
+
+    if (__same_memory_page(__user_ptr, __header) && __header->_M_uniq_const == __uniq_type_const)
+    {
+        __res = __header->_M_requested_number_of_bytes;
+    }
+    else
+    {
+        __res = __original_msize(__user_ptr);
+    }
+    return __res;
+}
+
+#if _WIN64
+
+static std::size_t
+__internal_aligned_msize(void* __user_ptr, std::size_t __alignment, std::size_t __offset)
+{
+    if (__user_ptr == nullptr || !__is_power_of_two(__alignment))
+    {
+        errno = EINVAL;
+        return -1;
+    }
+
     std::size_t __res = 0;
-    if (__user_ptr != nullptr)
+    __block_header* __header = static_cast<__block_header*>(__user_ptr) - 1;
+
+    if (__same_memory_page(__user_ptr, __header) && __header->_M_uniq_const == __uniq_type_const)
+    {
+        __res = __header->_M_requested_number_of_bytes;
+    }
+    else
     {
+        __res = __original_aligned_msize_ptr(__user_ptr, __alignment, __offset);
+    }
+    return __res;
+}
 
+static void
+__internal_aligned_free(void* __user_ptr)
+{
+    __internal_free_param(__user_ptr, __original_aligned_free_ptr);
+}
+
+#if _DEBUG
+
+static void
+__internal_free_dbg(void* __user_ptr, int __type)
+{
+    if (__user_ptr != nullptr)
+    {
         __block_header* __header = static_cast<__block_header*>(__user_ptr) - 1;
 
         if (__same_memory_page(__user_ptr, __header) && __header->_M_uniq_const == __uniq_type_const)
         {
-            __res = __header->_M_requested_number_of_bytes;
+            __free_usm_pointer(__header);
         }
         else
         {
-            __res = __original_msize(__user_ptr);
+            __original_free_dbg_ptr(__user_ptr, __type);
         }
     }
-    return __res;
 }
 
+#endif // _DEBUG
+
+void*
+__original_malloc(std::size_t size)
+{
+    return malloc(size);
+}
+
+void*
+__original_aligned_alloc(std::size_t size, std::size_t alignment)
+{
+    return _aligned_malloc(size, alignment);
+}
+
+void*
+__original_realloc(void* __user_ptr, std::size_t __new_size)
+{
+    return __original_realloc_ptr(__user_ptr, __new_size);
+}
+
+void*
+__original_aligned_realloc(void* __user_ptr, std::size_t __new_size, std::size_t __new_alignment)
+{
+    return __original_aligned_realloc_ptr(__user_ptr, __new_size, __new_alignment);
+}
+
+static void*
+__internal_expand(void* /*__user_ptr*/, std::size_t /*__size*/)
+{
+    // do not support _expand()
+    return nullptr;
+}
+
+std::size_t
+__get_page_size()
+{
+    SYSTEM_INFO __si;
+    GetSystemInfo(&__si);
+    return __si.dwPageSize;
```

Comment:
How expensive is to call `GetSystemInfo` any time `__get_page_size()` is called?
I guess we can consider a static wrapper if we believe it is better:

```
struct __system_info
{
    SYSTEM_INFO _M_si;
    __system_info()
    {
        GetSystemInfo(&_M_si);
    }
};

std::size_t __get_page_size()
{
    static __system_info __info;
    return __info._M_si.dwPageSize;
}
```

What do you think?

Reply:
One is not cheap, let's do it despite it's covered by `static` on per-TU level.